src/main/java/net/minecraftforge/common/MapGenStructureManager.java

```diff
@@ -0,0 +1,184 @@
+            addStructureName(name);
+        }
+        structureMap = builder.build();
+    }
+
+    /**
+     * Add a structure name to the locate command autocomplete list.
+     * <p>
+     * Is automatically done on construction
+     *
+     * @param structure Name of the structure
+     */
+    protected static void addStructureName(String structure)
+    {
+            allStructureNames.add(structure);
+    }
+
+    /**
+     * @return A new array containing both the given strings as well as the registered structure names
+     */
+    public static String[] appendStructureNames(String[] vanilla)
+    {
+        return ArrayUtils.addAll(vanilla, allStructureNames.toArray(new String[0]));
+    }
+
+    /**
+     * Get the nearest structure with the given name to the given world position.
+     * Returns null if the structure is not found
+     */
+    @Nullable
+    public BlockPos getNearestStructurePos(World worldIn, String structureName, BlockPos position, boolean findUnexplored)
+    {
+        if (structureMap.containsKey(structureName))
+        {
+            return structureMap.get(structureName).getNearestStructurePos(worldIn, position, findUnexplored);
+        }
+        return null;
+    }
+
+    /**
+     * Check if the given world position is within a structure with the given name.
+     * Returns false if the structure is not found
+     */
+    public boolean isInsideStructure(World worldIn, String structureName, BlockPos pos)
+    {
+        if (structureMap.containsKey(structureName))
+        {
+            return structureMap.get(structureName).isInsideStructure(pos);
+        }
+        return false;
+    }
+
+    /**
+     * Get the creatures of the given type the structure at the given position would like to spawn.
+     *
+     * @param defaults List of spawn entries. Should contain the (biome's) default entries. Won't be modified
+     */
+    @Nonnull
+    public List<Biome.SpawnListEntry> getSpawns(List<Biome.SpawnListEntry> defaults, World world, EnumCreatureType creatureType, BlockPos pos)
+    {
+        for (MapGenStructure structure : structureMap.values())
+        {
+            if (structure instanceof ISpawningStructure && structure.isPositionInStructure(world,pos))
+            {
+                return ((ISpawningStructure) structure).getSpawns(Collections.unmodifiableList(defaults), world, creatureType, pos);//Ensure mods don't accidentally modify the biomes spawn list itself
+            }
+        }
+        return defaults;
+    }
+
+    public void generateStructure(World world, Random rand, ChunkPos chunkPos)
+    {
+        for (MapGenStructure structure : structureMap.values())
+        {
+            structure.generateStructure(world, rand, chunkPos);
+        }
+    }
+
+    public boolean retroGenerateStructures(World world, Random rand, ChunkPos chunkPos, Chunk chunk){
+        boolean flag=false;
+        for(MapGenStructure structure : structureMap.values()){
```

Comment:
Please fix formatting in this method, it's all strange.

Reply:
Sorry about that, I keep forgetting that I disabled the auto-formatter for Forge